### PR TITLE
Fix CSL rendering in case of article

### LINKED ISSUE #8372: Entry preview not rendering the citation properly
### JabRef version

5.4 (latest release)

### Operating system

Windows

### Details on version and operating system

Windows 11 21H2

### Checked with the latest development build

- [X] I made a backup of my libraries before testing the latest development version.
- [X] I have tested the latest development version and the problem persists

### Steps to reproduce the behaviour

1. Go to Options > Preferences > Entry Preview.
2. Add "American Psychological Association 7th Edition" or "American Psychological Association 6th Edition".
3. Click `Save`.
4. Open a library containing an article with both `volume` and `number` fields. 
5. Double click the entry to see the `Entry Preview`. Check if entry preview is set to the "American Psychological Association 7th Edition" style guide.

You will notice that instead of showing the entry as "Smith, B., Jones, B., & Williams, J. (2016). Title of the test entry. _BibTeX Journal, 34_(3), 45–67. https://doi.org/10.1001/bla.blubb", the Entry Preview shows it as "Smith, B., Jones, B., & Williams, J. (2016). Title of the test entry. _BibTeX Journal, 34_, 45–67. https://doi.org/10.1001/bla.blubb" (i.e. without the `number` inside parenthesis). In the `Options` menu, however, the entry rendering appears to follow the style guide.

Until version 5.3, the entry preview was showing as expected. The issue appears only in versions 5.4 and 5.5 (so far).

### Appendix

BibTex source: 

```bibtex
@Article{,
  author  = {Smith, B. and Jones, B. and Williams, J.},
  journal = {BibTeX Journal},
  title   = {Title of the test entry},
  year    = {2016},
  number  = {3},
  pages   = {45--67},
  volume  = {34},
  doi     = {bla.blubb},
}
```

Screenshot of error:

### PR DESCRIPTION
**Pull request by koppor**
Fixes #8372
Fixes https://github.com/koppor/jabref/issues/514
Fixes https://discourse.jabref.org/t/unable-to-edit-my-bibtex-file-that-i-used-before-vers-5-1/2390/4
Fixes https://discourse.jabref.org/t/jabref-5-6-need-help-with-export-from-jabref-to-microsoft-word-entry-preview-of-apa-7-not-rendering-correctly/3462/6

Currently, only test cases showing the intended output are added. We still need to discuss where to put the translation logic. Since the CSL processor is not aware of bibtex/biblatex, we should do the translation based on the context. The translation logic is shown at https://github.com/JabRef/jabref/issues/8372#issuecomment-1014941935. The horizontal arrows mean: put at that place if not existant. A similar translationlogic (excel table style) is shown in https://github.com/JabRef/jabref/pull/8607#issuecomment-1210407418

Next steps:

- [ ] Implement translation logic
- [ ] Check if test cases match.
- [ ] If tests do not match, think of the edge cases and fix them

---

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by Siedlerchr** [comment]
The logic for "translating" would be needed here: A while back I already passed down the bibdatabasecontext so we can get the mode.:
https://github.com/JabRef/jabref/blob/b16160e5ab0fd9dc3aefc6824f55739b40d3a986/src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java#L97-L126

**Comment by koppor** [comment]
Current state: At test 13, we pass both issue and number to CSL Proc, but not a "page", it is not displayed.

**Review by ThiloteE — DISMISSED** [review]

**Comment by ThiloteE** [comment]
I think this fixes also https://discourse.jabref.org/t/unable-to-edit-my-bibtex-file-that-i-used-before-vers-5-1/2390/4 and https://github.com/koppor/jabref/issues/514

File: src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java [hunk 1]
```diff
@@ -108,8 +117,46 @@ private static CSLItemData bibEntryToCSLItemData(BibEntry bibEntry, BibDatabaseC
...
+                        bibEntry.clearField(StandardField.NUMBER);
+                    });
+                    bibEntry.getField(StandardField.PAGES).ifPresent(pages -> {
+                        if (pages.toLowerCase(Locale.ROOT).startsWith("article ")) {
+                            pages = pages.substring("Article ".length());
+                            bibEntry.setField(StandardField.NUMBER, pages);
+                        }
+                    });
+                    bibEntry.getField(StandardField.EID).ifPresent(eid -> {
+                        if (!bibEntry.hasField(StandardField.PAGES)) {
```
Comment (ThiloteE) [inline src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java:150]: `hasField` here, might lead to tests failing because if `eid` is present, but `pages` field is NOT present, eid will not get rendered, right?

Comment (ThiloteE) [reply src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java:150]: The more I think about it: Some tests might fail, simply, because we make the preview render `eid` for bibtex mode. Since eid is not part of the bibtex standard, I came up with tests that basically ignore eid, in bibtex mode and only take eid into account if JabRef users use BibLaTeX mode. So the tests would need to be changed if we include lines 149 - 154. We could remove these lines altogether and see how tests play out without.

File: src/test/java/org/jabref/logic/citationstyle/CitationStyleGeneratorTest.java [hunk 2]
```diff
@@ -166,85 +170,427 @@ void testHandleCrossRefFields() {
...
+                                .withField(StandardField.AUTHOR, "Last, First and Doe, Jane")
+                                .withField(StandardField.NUMBER, "28"),
+                        "ieee.csl"),
+                Arguments.of(
+                        "[1]F. Last and J. Doe, no. 7, Art. no. 28.\n",
+                        BibDatabaseMode.BIBLATEX,
+                        new BibEntry(StandardEntryType.Article)
+                                .withField(StandardField.AUTHOR, "Last, First and\nDoe, Jane")
+                                .withField(StandardField.ISSUE, "7")
+                                .withField(StandardField.NUMBER, "28"),
```
Comment (ThiloteE) [inline src/test/java/org/jabref/logic/citationstyle/CitationStyleGeneratorTest.java:203]: Replace `.withField(StandardField.NUMBER, "28"),` with `.withField(StandardField.EID, "55"),`

Reason: "Art. no. xx" is only necessary for the EID field. See page "II. Style—13" at https://journals.ieeeauthorcenter.ieee.org/wp-content/uploads/sites/7/IEEE_Reference_Guide.pdf

File: src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java [hunk 3]
```diff
@@ -108,8 +117,46 @@ private static CSLItemData bibEntryToCSLItemData(BibEntry bibEntry, BibDatabaseC
...
+                        bibEntry.setField(StandardField.NUMBER, eid);
+                        bibEntry.clearField(StandardField.EID);
+                    });
+                } else {
+                    // BibTeX mode
+                    bibEntry.getField(StandardField.NUMBER).ifPresent(number -> {
+                        bibEntry.setField(StandardField.ISSUE, number);
+                        bibEntry.clearField(StandardField.NUMBER);
+                    });
+                    bibEntry.getField(StandardField.PAGES).ifPresent(pages -> {
```
Comment (ThiloteE) [inline src/main/java/org/jabref/logic/citationstyle/CSLAdapter.java:143]: I think the order could be wrong here. Maybe move lines 143 - 148 behind 154.

**Comment by Siedlerchr** [comment]
Let's create a spreadsheet /matrix with bibtex fields and their mappings and biblatex fields then we can better check the test cases. And a condition column

**Comment by ThiloteE** [comment]
|BibTeX|BibLaTeX|EntryPreview/CSL| extra logic, conditions and info |
|:---|:---|:---|:----|
| volume | volume | volume | |
| number | issue     | issue | For conversion to CSL or BibTeX: BibLaTeX `number` > BibLaTeX `issue`|
| number | number | issue | For conversion to CSL or BibTeX: BibLaTeX `number` > BibLaTeX `issue`|
| pages | eid | number | Some journals put the article-number into the pages field. If BibLaTeX `eid` exists, use csl `number`. If `pages` exists, use csl `page`. If `eid` WITHIN the `pages` field exists, detect the eid and use csl `number`. If both `eid` and `pages` exists, ideally use both csl `number` and csl `page`. Ideally the citationstyle should be able to flexibly choose the rendering. If this is not possible then use csl `page` only, since this will comply with APA 7th edition: If both csl `number` and csl `page` exist, then only take csl `page`|
| pages | pages | page | Some journals put the article-number into the pages field. If BibLaTeX `eid` exists, use csl `number`. If `pages` exists, use csl `page`.  If `eid` WITHIN the `pages` field exists, detect the eid and use csl `number`. If both `eid` and `pages` exists, ideally use both csl `number` and csl `page`. Ideally the citationstyle should be able to flexibly choose the rendering. If this is not possible then use csl `page` only, since this will comply with APA 7th edition: If both csl `number` and csl `page` exist, then only take csl `page` |

**Comment by Siedlerchr** [comment]
TIL: Apa CSL has no field `number`

**Comment by ThiloteE** [comment]
Known problems of this implementation: A negative tradeoff between rendering IEEE or APA style correctly. Until APA 6th edition CSL style is able to render "number", I would suggest to map eid to the page field, otherwise eid will not get rendered at all. Unfortunately this means that IEEE now renders article numbers in the pages field and in the eid field with `p.` prefixed. Overall, I would suggest this to be an improvement to the status quo.

**Comment by ThiloteE** [comment]
Waiting for external CSL maintainers a little bit. We have time until JabRef 5.8 will be released anyway....

**Review by Siedlerchr — APPROVED** [review]

**Review by ThiloteE — APPROVED** [review]
